PFW-1503 Improve live Z babystep menu handling

gudnimg opened the pull request:
Short history:
* I broke the behavior in here https://github.com/gudnimg/Prusa-Firmware/commit/2276217cbb47501a0558d988446c8d1a863a3384 with optimisations. This created a bug where the menu would close and not save the last Z value into EEPROM.
* The bug above was fixed recently here https://github.com/prusa3d/Prusa-Firmware/pull/4005 by @3d-gussner but later we discovered this created another bug where `menu_leave` is set to `1`, but it is then never set to `0` again. This bug appears only when first layer calibration is run twice consequatively and the LCD will freeze or stop behaving correctly.

In this PR I have a proposed solution which fixes the reported problem:

Improvements
====

* Add `BABYSTEP_ALLOWED` function to unify checks whether or not babystepping is allowed, if not the menu is dismissed and hidden from the user.
* The Z baby step menu closes automatically when first layer calibration is done. This is expected behavior but additionally now we don't need to wait for a timeout or close the menu manually by setting a variable.
* Do not allow the Z baby step menu to timeout when LcdCommands != Idle. This covers first layer calibration. In all other situations it times out as normal (after 90 seconds).

Change in memory:
Flash: -182 bytes
SRAM: 0 bytes

gudnimg committed:
PFW-1503 Improve live Z babystep menu handling

Fixes an issue with running first layer calibration twice in a row.

Improvements:
Now the Z baby step menu closes automatically when first layer calibration is done.
No need to wait for a timeout or close the menu manually by setting a variable

If the baby stepping menu is open, and suddenly
the printer enters a state where baby stepping
is not allowed.
The printer will save the last value before closing the menu.

When LcdCommands != Idle, don't dismiss
the Z baby step menu. This saves 20B

Change in memory:
Flash: -130 bytes
SRAM: 0 bytes

gudnimg commented:
I suspect my `blocks_queued()` condition is not sufficient alone. Sometimes the screen is being dismissed. Need to think about how to add more to this check without increasing the code size too much.

gudnimg committed:
PFW-1503 Add additional check

Add lcd_commands_type == LcdCommands::Layer1Cal
just in case blocks_queued() is 0 for one instant between
lcd_command steps

Change in memory:
Flash: +8 bytes
SRAM: 0 bytes

DRracer commented on Firmware/Marlin.h:
```diff
@@ -365,6 +365,13 @@ extern bool printer_active();
+//! 2) Not allowed during Homing (printer busy)
+//! 3) Not allowed during Mesh Bed Leveling (printer busy)
+//! 4) Allowed if there are queued blocks OR there is a print job running
+bool BABYSTEP_ALLOWED();
```
/nitpick/: I'd avoid naming the function in uppercase letters

gudnimg replied:
Fixed in [65e9e86](https://github.com/prusa3d/Prusa-Firmware/pull/4063/commits/65e9e864c0eb8b1fd9bdac9ac549ed4b981e1976)

DRracer commented on Firmware/ultralcd.cpp:
```diff
@@ -5342,7 +5332,7 @@ static void lcd_main_menu()
 #endif //TMC2130_DEBUG
 
-    if ( ( IS_SD_PRINTING || usb_timer.running() || (lcd_commands_type == LcdCommands::Layer1Cal)) && (current_position[Z_AXIS] < Z_HEIGHT_HIDE_LIVE_ADJUST_MENU) && !homing_flag && !mesh_bed_leveling_flag) {
+    if ( BABYSTEP_ALLOWED() ) {
```
beware: the `current_position[Z_AXIS] < Z_HEIGHT_HIDE_LIVE_ADJUST_MENU` is important. The check was not about the known positions of Z, but about allowing to enter the live-z menu during printing of the first few layers (and not allow any later)

gudnimg replied:
Github marks this as outdated due to me rebasing the branch to sync with MK3. This comment is still relevant 👍

gudnimg replied:
[e3b9e9e](https://github.com/prusa3d/Prusa-Firmware/pull/4063/commits/e3b9e9efe0aeb9361faa7a7e93c877f250c540fe) This is now applied everywhere so it's consistant in behavior.

DRracer commented on Firmware/ultralcd.cpp:
```diff
@@ -7457,13 +7447,11 @@ void menu_lcd_longpress_func(void)
-    if ( (moves_planned() || IS_SD_PRINTING || usb_timer.running() )){ // long press as live-adj-z
-        if (( current_position[Z_AXIS] < Z_HEIGHT_HIDE_LIVE_ADJUST_MENU ) // only allow live-adj-z up to 2mm of print height
-        && ( menu_menu == lcd_status_screen // and in listed menus...
+    if (BABYSTEP_ALLOWED()){ // long press as live-adj-z
```
oh, this is the primary spot for the `current_position[Z_AXIS] < Z_HEIGHT_HIDE_LIVE_ADJUST_MENU` check

gudnimg replied:
Github marks this as outdated due to me rebasing the branch to sync with MK3. This comment is still relevant 👍

gudnimg replied:
[e3b9e9e](https://github.com/prusa3d/Prusa-Firmware/pull/4063/commits/e3b9e9efe0aeb9361faa7a7e93c877f250c540fe) This is now applied everywhere so it's consistant in behavior.

gudnimg committed:
PFW-1503 Change BABYSTEP_ALLOWED to babystep_allowed

I initially started with a preprocessor macro
called  BABYSTEP_ALLOWED, but now it is a function.
It more common to use lower case for function names

gudnimg committed:
PFW-1503 only allow babystepping on the first couple of layers

Change in memory:
Flash: +28 bytes
SRAM: 0 bytes

gudnimg committed:
PFW-1503 drop {} for consistency

DRracer reviewed (approved):
Even though the conditions are a bit different compared to the previous implementation, it looks acceptable to me :+1: 

We should gather more feedback during testing

3d-gussner reviewed (changes requested):
On MK404 during LCD -> Calibration -> First layer cal. it jumps out quite fast to status screen, while it should stay in the Live Z-adjust menu until it is finished.

gudnimg commented:
> On MK404 during LCD -> Calibration -> First layer cal. it jumps out quite fast to status screen, while it should stay in the Live Z-adjust menu until it is finished.

Ah I think it is this condition kicking in: `current_position[Z_AXIS] < Z_HEIGHT_HIDE_LIVE_ADJUST_MENU)` I saw this before but forgot about it when I added the condition back. 

The extruder lifts up above 2mm while moving to the start position. 

Unfortuately I can’t fix it until this evening (after work). Feel free to add a patch if possible.

gudnimg committed:
Implement printJobOngoing()

This makes the if statments a little bit more readable

Change in memory:
Flash: -88 bytes
SRAM: 0 bytes

gudnimg committed:
PFW-1503 Fix an issue where the menu is dismissed in First layer calibraiton

When the extruder lifts up after completing the Purge line,
 the baby stepping is not allowed for a short time. This dismisses
 the menu. We don't want this behavior, so only apply the Z-axis requirement
 when printing.

 Change in memory:
 Flash: +8 bytes
 SRAM: 0 bytes

gudnimg commented:
@3d-gussner I've fixed the issue, please re-test when you have time :)

gudnimg committed:
Header cleanup

3d-gussner reviewed (approved):
Tested on MK404
